core/queryalgebra/evaluation/src/main/java/org/eclipse/rdf4j/query/algebra/evaluation/impl/DefaultEvaluationStrategy.java

```diff
@@ -710,10 +709,26 @@ protected QueryEvaluationStep prepare(DescribeOperator node, QueryEvaluationCont
 	protected QueryEvaluationStep prepare(Distinct node, QueryEvaluationContext context)
 			throws QueryEvaluationException {
 		final QueryEvaluationStep child = precompile(node.getArg(), context);
-		return bindings -> {
-			final CloseableIteration<BindingSet> evaluate = child.evaluate(bindings);
-			return new DistinctIteration<BindingSet>(evaluate,
-					DefaultEvaluationStrategy.this::makeSet);
+		final CollectionFactory cf = this.getCollectionFactory().get();
+		return new QueryEvaluationStep() {
+
+			@Override
+			public CloseableIteration<BindingSet> evaluate(BindingSet bindings) {
+				final CloseableIteration<BindingSet> evaluate = child.evaluate(bindings);
+				return new DistinctIteration<BindingSet>(evaluate,
+						cf.createSetOfBindingSets()) {
+
+					@Override
+					protected void handleClose() throws QueryEvaluationException {
+						try {
+							cf.close();
+						} catch (QueryEvaluationException e) {
+							super.handleClose();
```

Comment:
Should this maybe be try-finally or is super.handleClose() only supposed to be called when there is an exception?

Reply:
This should be a try-finally, although right now super.handleClose is a noop. That could change in the future.
will-fix.